src/Microsoft.Health.Fhir.SqlServer/Features/Storage/SqlServerFhirDataStore.cs

```diff
@@ -0,0 +1,243 @@
+                            }
+                            catch (SqlException e)
+                            {
+                                switch (e.Number)
+                                {
+                                    case SqlErrorCodes.NotFound:
+                                        throw new MethodNotAllowedException(Resources.ResourceCreationNotAllowed);
+                                    case SqlErrorCodes.PreconditionFailed:
+                                        throw new ResourceConflictException(weakETag);
+                                    default:
+                                        _logger.LogError(e, "Error from SQL database on upsert");
+                                        throw;
+                                }
+                            }
+                        }
+                    }
+                }
+            }
+        }
+
+        public async Task<ResourceWrapper> GetAsync(ResourceKey key, CancellationToken cancellationToken)
+        {
+            using (var connection = new SqlConnection(_configuration.ConnectionString))
+            {
+                await connection.OpenAsync(cancellationToken);
+
+                using (SqlCommand command = connection.CreateCommand())
+                {
+                    command.Parameters.AddWithValue("@resourceTypeId", _model.GetResourceTypeId(key.ResourceType));
+                    command.Parameters.AddWithValue("@resourceId", key.Id);
+
+                    bool versionedRead = !string.IsNullOrEmpty(key.VersionId);
+
+                    if (versionedRead)
+                    {
+                        command.CommandText = @"
+                            SELECT Version, LastUpdated, IsDeleted, IsHistory, RawResource 
+                            FROM dbo.Resource
+                            WHERE ResourceTypeId = @resourceTypeId AND ResourceId = @resourceId AND Version = @version";
+
+                        if (!int.TryParse(key.VersionId, out var version))
+                        {
+                            return null;
+                        }
+
+                        command.Parameters.AddWithValue("@version", version);
+                    }
+                    else
+                    {
+                        command.CommandText = @"
+                            SELECT Version, LastUpdated, IsDeleted, IsHistory, RawResource 
+                            FROM dbo.Resource
+                            WHERE ResourceTypeId = @resourceTypeId AND ResourceId = @resourceId AND IsHistory = 0";
+                    }
+
+                    using (SqlDataReader sqlDataReader = await command.ExecuteReaderAsync(CommandBehavior.SequentialAccess, cancellationToken))
+                    {
+                        if (!sqlDataReader.Read())
+                        {
+                            return null;
+                        }
+
+                        int version = sqlDataReader.GetInt32(0);
```

Comment:
What about getting the values by column name (e.g. `sqlDataReader["Version"]`)? Alternatively, I think we should use a string variable in the above queries to represent the select and from lines and only vary the where. Right now, it seems like it would be easy for them to get out of sync.

Reply:
Added debug check extension methods. `sqlDataReader.GetInt32("fieldname", 0)`